src/index.ts

```diff
@@ -377,6 +412,155 @@ export class Upload extends Pumpify {
     });
   }
 
+  /** A stream representing the incoming data to upload */
+  private readonly upstream = new Duplex({
+    read: async () => {
+      this.once('prepareFinish', () => {
+        // Allows this (`Upload`) to finish/end once the upload has succeeded.
+        this.upstream.push(null);
+      });
+    },
+    write: this.writeToChunkBuffer.bind(this),
+  });
+
+  /**
+   * A handler for `upstream` to write and buffer its data.
+   *
+   * @param chunk The chunk to append to the buffer
+   * @param encoding The encoding of the chunk
+   * @param readCallback A callback for when the buffer has been read downstream
+   */
+  private writeToChunkBuffer(
+    chunk: Buffer | string,
+    encoding: BufferEncoding,
+    readCallback: () => void
+  ) {
+    this.upstreamChunkBuffer = Buffer.concat([
+      this.upstreamChunkBuffer,
+      typeof chunk === 'string' ? Buffer.from(chunk, encoding) : chunk,
+    ]);
+    this.chunkBufferEncoding = encoding;
+
+    this.once('readFromChunkBuffer', readCallback);
+    process.nextTick(() => this.emit('wroteToChunkBuffer'));
+  }
+
+  /**
+   * Prepends data back to the upstream chunk buffer.
+   *
+   * @param chunk The data to prepend
+   */
+  private unshiftChunkBuffer(chunk: Buffer) {
+    this.upstreamChunkBuffer = Buffer.concat([chunk, this.upstreamChunkBuffer]);
+  }
+
+  /**
+   * Retrieves data from upstream's buffer.
+   *
+   * @param limit The maximum amount to return from the buffer.
+   * @returns The data requested.
+   */
+  private pullFromChunkBuffer(limit: number) {
+    const chunk = this.upstreamChunkBuffer.slice(0, limit);
+    this.upstreamChunkBuffer = this.upstreamChunkBuffer.slice(limit);
+
+    // notify upstream we've read from the buffer so it can potentially
+    // send more data down.
+    process.nextTick(() => this.emit('readFromChunkBuffer'));
```

Comment:
Why does this need to occur on `nextTick`?

Reply:
Slight optimization - gives the opportunity for the chunk to be passed downstream before requesting more data to be immediately read